event_handlers.go

```diff
@@ -0,0 +1,100 @@
+package lightstep
+
+import (
+	"log"
+	"sync"
+	"sync/atomic"
+)
+
+func init() {
+	OnEvent(NewOnEventLogOneError())
+}
+
+var eventHandler atomic.Value
+var eventHandlerLock sync.Mutex
+
+// An OnEventHandler can be registered with OnEvent to
+type OnEventHandler func(Event)
+
+// onEvent is a thread-safe function for emiting tracer events.
+func onEvent(event Event) {
+	handler := eventHandler.Load().(OnEventHandler)
+	handler(event)
+}
+
+// OnEvent sets a global handler to receive tracer events as they occur. Events
+// may be emitted by the tracer, or by calls to static functions in this package.
+// It is suggested that you set your OnEvent handler before starting your tracer,
+// but it is safe to set a new handler at any time. Setting a new handler removes
+// the previous one.
+//
+// OnEvent is called synchronously – do not block in your handler as it may interfere
+// with the tracer. If no OnEvent handler is provided, a LogOnceOnError handler
+// is used by default.
+//
+// NOTE: Event handling is for reporting purposes only. It is not intended as a
+// mechanism for controling or restarting the tracer. Connection issues, retry
+// logic, and other transient errors are handled internally by the tracer.
+func OnEvent(handler OnEventHandler) {
```

Comment:
The fact that this can only be called once makes me think it should be provided to the NewTracer constructor. Otherwise, any rogue library could install its own handler. This is something I believe should be set in main, why make it possible to change?

Reply:
@jmacd please see the part of the PR description called "Why a global OnEvent handler?" for my reasoning.

Reply:
There are two questions there, one "Why" and one "Why not". There's only one answer for the Why I can see:

_Events may occur inside of package-level functions, or deep within tracer subsystems. A global handler allows these constructs to emit events without the need to thread a reporter into hard to get areas._

This to me is suspect. We already have a global tracer, and if that global tracer is a lightstep tracer, then we can instead get the global tracer, see if it's a lightstep tracer, then cast it, get the event handler, and use it.

Having the event handler be registered by a global function just seems to invite misuse. Registering it when the tracer starts feels less prone to surprise.

Reply:
I tend to agree. It seems like having a global is asking for action at a distance and confusing outcomes.

Reply:
Totally understand the critique! I too, hate globals and find them suspicious. It is possible to move event handling back to If we want to allow event reporting be an `Option` on a Tracer instead of a global handler, there are three things we have to do:

- cease to emit events for typecast errors in `lightstep.Flush` and `lightstep.Close`.
- refactor `collectorClient` implementations to take an event handler.
- live with the oddness around using an invalid Options object to report startup failures.

The second two are doable. Ceasing to emit events though... I'm concerned about that. That gets at the core design requirements of this feature, which is to report diagnostics while no longer returning errors from lightstep functions.

An example of this situation btw, just so it's clear that it's worth thinking about: a customer instruments all of their code with LightStep. But they they read a cool OpenTracing article and wrap the lightstep tracer up in some kind of meta tracer. Then their flushing and shutdown code ceases to function. That means dropped spans. When we silently dropped spans on shutdown in the past, users were pretty mad when they eventually noticed.

```go
func startTracer(){
  lsTracer := lightstep.NewTracer(opts)

  // when they wrap our tracer up
  logger := glog.NewLogger()
  tracer := ot_helpers.NewLogTeeTracer(lsTracer, logger)

  // and register the wrapped tracer...
  opentracing.SetGlobalTracer(tracer)
}

func stopTracer(ctx context.Context){
  // ...shutdown will start failing here. Is this a silent failure? How do we notify?
  lightstep.Close(ctx, opentracing.GlobalTracer())
}
```

A relevant design input is that users have reacted poorly to there being `error` return values on lightstep functions. When users see an error, they want to "handle" it somehow... it's been difficult to explain to users that they should alert on errors but otherwise ignore them. It's been enough of a problem that one of the goals was to unify the reporting mechanism in such a way that it would not encourage users to try to create a feedback loop with the tracer. For example, receiving a flush error and trying to manually close and restart the tracer, rather than letting the tracer handle it. Patterns like that have lead to users writing code that blocks shutdown, and other problems.

The alternative solution of grabbing at the global tracer to get an event handler won't work in the situation above. I would suggest we go with an explicit global event handler, or a solution for the above issues which does not require any globals.

Hope that helps!

Reply:
I stand convinced. Ted's explanation revolves around corner cases that users legitimately fall into, like retrying a lightstep call that would fail and never notify the user (because a different tracer is installed) if we didn't make the event notifier global.